### PR TITLE
Optimize `Func::call` and its C API

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an alternative to #3298 which achieves effectively the
same goal of optimizing the `Func::call` API as well as its C API
sibling of `wasmtime_func_call`. The strategy taken here is different
than #3298 though where a new API isn't created, rather a small tweak to
an existing API is done. Specifically this commit handles the major
sources of slowness with `Func::call` with:

* Looking up the type of a function, to typecheck the arguments with and
  use to guide how the results should be loaded, no longer hits the
  rwlock in the `Engine` but instead each `Func` contains its own
  `FuncType`. This can be an unnecessary allocation for funcs not used
  with `Func::call`, so this is a downside of this implementation
  relative to #3298. A mitigating factor, though, is that instance
  exports are loaded lazily into the `Store` and in theory not too many
  funcs are active in the store as `Func` objects.

* Temporary storage is amortized with a long-lived `Vec` in the `Store`
  rather than allocating a new vector on each call. This is basically
  the same strategy as #3294 only applied to different types in
  different places. Specifically `wasmtime::Store` now retains a
  `Vec<u128>` for `Func::call`, and the C API retains a `Vec<Val>` for
  calling `Func::call`.

* Finally, an API breaking change is made to `Func::call` and its type
  signature (as well as `Func::call_async`). Instead of returning
  `Box<[Val]>` as it did before this function now takes a
  `results: &mut [Val]` parameter. This allows the caller to manage the
  allocation and we can amortize-remove it in `wasmtime_func_call` by
  using space after the parameters in the `Vec<Val>` we're passing in.
  This change is naturally a breaking change and we'll want to consider
  it carefully, but mitigating factors are that most embeddings are
  likely using `TypedFunc::call` instead and this signature taking a
  mutable slice better aligns with `Func::new` which receives a mutable
  slice for the results.

Overall this change, in the benchmark of "call a nop function from the C
API" is not quite as good as #3298. It's still a bit slower, on the
order of 15ns, because there's lots of capacity checks around vectors
and the type checks are slightly less optimized than before. Overall
though this is still significantly better than today because allocations
and the rwlock to acquire the type information are both avoided. I
personally feel that this change is the best to do because it has less
of an API impact than #3298.

### CODE REVIEW

**Comment by alexcrichton** [comment]
To show an idea of the numbers I'm looking at, here's some information. I've got a [C program](https://gist.github.com/alexcrichton/e3ad76042c847981ab064e59325635aa) that shows the following timings on various wasmtime versions. Timings are shown with using `wasm.h` and `wasmtime.h`. Times are all in nanoseconds.

What's being benchmarked:

## nop

**host calls a wasm function that does nothing**

| version | `wasm.h` |  `wasmtime.h` |
|----------|--------------|-----------------|
|  0.26.1 | 45 | |
| 0.27.0 | 104 | |
| 0.28.0 | 115 |  113 |
| 0.29.0 | 117 | 117 |
| this PR | 43 | 51 |


## i64 
**host calls a wasm function. Wasm function calls a host function returning i64. Wasm then returns that i64.**

| version | `wasm.h` |  `wasmtime.h` |
|----------|--------------|-----------------|
|  0.26.1 | 181 | |
| 0.27.0 | 247 | |
| 0.28.0 | 258 |  240 |
| 0.29.0 | 258 | 245 |
| this PR | 130 | 104 |


## many 
**host calls a wasm function. Wasm function calls a host function with 5 i32 params and one i32 return. Wasm discards result and returns.**



| version | `wasm.h` |  `wasmtime.h` |
|----------|--------------|-----------------|
|  0.26.1 | 223 | |
| 0.27.0 | 275 | |
| 0.28.0 | 295 |  258 |
| 0.29.0 | 284 | 261 |
| this PR | 173 | 138 |

---

This is all a far cry away from what Rust can do though, so while this PR certainly improves the state of affairs Rust [far outstrips it](https://gist.github.com/alexcrichton/59590c6509790cace6bef5b9d15ba69d). Rust has the option of using "typed" and "untyped" functions, effectively using `Func` or `TypedFunc` to call wasm functions or `Func::wrap`/`Func::new` to define host functions. The table below shows timings (in nanoseconds) of calling the functions in a typed/untyped fashion:

**when imports are defined with `Func::wrap`**

| func | typed | untyped |
|-----|-------|---------|
| nop | 20 | 34 |
| i64 | 21 | 41 |
| many | 23 | 37 |

**when imports are defined with `Func::new`**

| func | typed | untyped |
|-----|-------|---------|
| i64 | 43 | 62 |
| many | 85 | 100 |

The basic summary here is that when we have "typed" versions of everything basically any combination of arguments/returns is roughly 20ns of call-time overhead. That's the minimum threshold for calling a wasm function. Once "untyped" things are used then there starts to be per-argument and per-result overhead. This creeps up in both calling wasm functions and calling host functions, and can be seen how "many" is slower than "i64" which is slower than "nop". Finally the C API imposes about a 40ns blanket overhead on top of the wasmtime "untyped" Rust versions, mostly because `wasmtime_val_t` needs to be converted to `Val` and back.

Basically at this point I think that the "untyped" variants, which C currently is forced to use, are as optimal as they're gonna get. There's still inherent overhead with translation between `wasmtime_val_t` and `Val`, but at some point that can only be but so optimal. I think the only way forward to improve C bindings is to somehow get a "typed" variant in there which opens up opportunities to remove type translations and type checks entirely.

**Comment by alexcrichton** [comment]
I'll note that I think we still want this even in the face of https://github.com/bytecodealliance/wasmtime/pull/3345. I personally like the idea of taking `&mut [Val]` instead of returning `Box<[Val]>` for `Func::call`, and otherwise all the optimizations here are internal implementation details we can fiddle with over time.

**Review by peterhuene — APPROVED** [review]
Looks great, just the one question!

File: crates/c-api/src/func.rs [hunk 1]
```diff
@@ -118,23 +133,20 @@ pub unsafe extern "C" fn wasm_func_call(
     let f = func.func();
     let results = (*results).as_uninit_slice();
     let args = (*args).as_slice();
-    if results.len() != f.ty(func.ext.store.context()).results().len() {
-        return Box::into_raw(Box::new(wasm_trap_t::new(
-            anyhow!("wrong number of results provided").into(),
-        )));
-    }
-    let params = args.iter().map(|i| i.val()).collect::<Vec<_>>();
+    let mut dst = Vec::new();
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/c-api/src/func.rs:136]: We could yeah. The reason I didn't do it here is that this is part of the `wasm.h` C API which uses different store data. We could update that too but I figured I didn't want to worry much about this and thought I could deal with it later if necessary. Given the later `*_unchecked` APIs I ended up adding I think that no one's gonna use these for the most performance hooks anyway.

Do you agree it's alright to leave this for some future refactoring, if necessary?

[the remark this answers is not in the document]
Comment (peterhuene) [reply crates/c-api/src/func.rs:136]: Definitely 👍
